pkg/config/conversion/runtime_conversion.go

```diff
@@ -0,0 +1,122 @@
+// SPDX-FileCopyrightText: 2024 The Crossplane Authors <https://crossplane.io>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package conversion
+
+import (
+	"slices"
+	"sort"
+	"strings"
+
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/pkg/errors"
+)
+
+// Mode denotes the mode of the runtime API conversion, e.g.,
+// conversion of embedded objects into singleton lists.
+type Mode int
+
+const (
+	// ToEmbeddedObject represents a runtime conversion from a singleton list
+	// to an embedded object, i.e., the runtime conversions needed while
+	// reading from the Terraform state and updating the CRD
+	// (for status, late-initialization, etc.)
+	ToEmbeddedObject Mode = iota
+	// ToSingletonList represents a runtime conversion from an embedded object
+	// to a singleton list, i.e., the runtime conversions needed while passing
+	// the configuration data to the underlying Terraform layer.
+	ToSingletonList
+)
+
+// String returns a string representation of the conversion mode.
+func (m Mode) String() string {
+	switch m {
+	case ToSingletonList:
+		return "toSingletonList"
+	case ToEmbeddedObject:
+		return "toEmbeddedObject"
+	default:
+		return "unknown"
+	}
+}
+
+// setValue sets the value, in pv, to v at the specified path fp.
+// It's implemented on top of the fieldpath library by accessing
+// the parent map in fp and directly setting v as a value in the
+// parent map. We don't use fieldpath.Paved.SetValue because the
+// JSON value validation performed by it potentially changes types.
+func setValue(pv *fieldpath.Paved, v any, fp string) error {
+	segments := strings.Split(fp, ".")
+	p := fp
+	var pm any = pv.UnstructuredContent()
+	var err error
+	if len(segments) > 1 {
+		p = strings.Join(segments[:len(segments)-1], ".")
+		pm, err = pv.GetValue(p)
+		if err != nil {
+			return errors.Wrapf(err, "cannot get the parent value at field path %s", p)
+		}
+	}
+	parent, ok := pm.(map[string]any)
+	if !ok {
+		return errors.Errorf("parent at field path %s must be a map[string]any", p)
+	}
+	parent[segments[len(segments)-1]] = v
+	return nil
+}
+
+// Convert performs conversion between singleton lists and embedded objects
+// while passing the CRD parameters to the Terraform layer and while reading
+// state from the Terraform layer at runtime. The paths where the conversion
+// will be performed are specified using paths and the conversion mode (whether
+// an embedded object will be converted into a singleton list or a singleton
+// list will be converted into an embedded object) is determined by the mode
+// parameter.
+func Convert(params map[string]any, paths []string, mode Mode) (map[string]any, error) { //nolint:gocyclo // easier to follow as a unit
+	switch mode {
+	case ToSingletonList:
+		slices.Sort(paths)
+	case ToEmbeddedObject:
+		sort.Slice(paths, func(i, j int) bool {
+			return paths[i] > paths[j]
+		})
+	}
```

Comment:
The relevant issue here is the fieldpath expressions use indices (e.g., `[*]`) for traversing lists. We generate the fieldpath expressions for the singleton lists that have been converted using the expressions that work on the original Terraform topology. And we need to convert from embedded objects to singleton lists (when passing data from the XP layer to the TF layer) and from singleton lists to embedded objects (when passing data from the TF layer to the XP layer) at runtime using these fieldpath expressions. If we do not process the field conversions in the correct order, the expressions may mismatch the temporary topology during a conversion. For example, let's assume a singleton list contains another nested singleton list, both of which we've converted and let's assume the corresponding fieldpath expressions for the conversion targets are `a.b[*]` and `a.b[*].c[*]`, respectively. When converting from embedded objects to singleton lists, converting `a.b` (which is an embedded object in the source) before its children relieves us from having to adjust the fieldpath expressions with the `a.b[*].` prefix because once `a.b` is converted into a singleton list, then the fieldpath expressions that target its children now match the (converted) topology. Similarly, when converting from  singleton lists to embedded objects at runtime (e.g., after reading the TF state), converting `a.b[*]` after `a.b[*].c[*]` relieves us from having to adjust the fieldpath expressions that target `a.b[*]`'s children. 

This is the motivation of processing the conversions in an order. We could as well adjust the fieldpath expressions to match the topology as we perform the conversions, but this looks simpler to me.